fix(diagnostic): allow virtual_{lines,text} cursor exclusivity

mwcz opened the pull request:
**Problem:**

virtual_text diagnostics are great when skimming a file, and virtual_lines are great when "zooming in" on a particular problem. Having both enabled results in duplicate diagnostics on-screen.

First reported in #33092, with this image as an example:
![image](https://github.com/user-attachments/assets/0c0b87b4-ae09-4b9b-8c15-7529958103b5)

**Solution:**

This PR expands the behavior of `current_line` for virtual_text and virtual_lines by making `virtual_text.current_line = false` distinct from `nil`.  If you set:

```lua
vim.diagnostic.config({
  virtual_text = { current_line = false },
  virtual_lines = { current_line = true },
})
```

With this configuration, virtual_text will be used to display diagnostics until the cursor reaches the same line, at which point they will be hidden and virtual_lines will take its place.

The situation in the image above now behaves like this:

![image](https://github.com/user-attachments/assets/5d5db34c-a1c8-4fb4-8259-214fca1f9d6b)

![image](https://github.com/user-attachments/assets/4c57d82f-06e6-4b03-b294-48716c3e4382)

Note to maintainers: this is my first PR to neovim and I haven't looked into writing tests for this yet but I will look into it if desired.
<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

mwcz committed:
fix(diagnostic): allow virtual_{lines,text} cursor exclusivity

Problem:

virtual_text diagnostics are great when skimming a file, and
virtual_lines are great when "zooming in" on a particular problem.
Having both enabled results in duplicate diagnostics on-screen.

Solution:

This PR expands the behavior of `current_line` for virtual_text and
virtual_lines by making `virtual_text.current_line = false` distinct
from `nil`.  If you set:

    vim.diagnostic.config({
      virtual_text = { current_line = false },
      virtual_lines = { current_line = true },
    })

With this configuration, virtual_text will be used to display
diagnostics until the cursor reaches the same line, at which point they
will be hidden and virtual_lines will take its place.

justinmk commented:
if this is truly needed, `current_line` should be a string option instead of a boolean.

alternatively, should this just be the default behavior of `current_line=true` ?

> Having both enabled results in duplicate diagnostics on-screen.

Both are enabled, so yeah...

gpanders commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1589,12 +1562,16 @@ M.handlers.underline = {
 
   for line, line_diagnostics in pairs(diagnostics) do
     local virt_texts = M._get_virt_text_chunks(line_diagnostics, opts)
+    local should_render_current_line = (opts.current_line == true and line == lnum)
```
If `opts.current_line` is `nil` this should also be true.

mwcz replied:
That case is handled by the last part of the condition: `or (opts == true or type(opts) == 'table')`

gpanders replied:
`opts` will always be a table here, so `opts == true` will always be false and `type(opts) == 'table'` will always be true.

I think the logic would be simplified here for testing whether or not the line should be skipped:

```lua
local skip = (opts.current_line == true and line ~= lnum) or (opts.current_line == false and line == lnum)
```

mwcz replied:
Ah, thanks, that helps a lot.  I saw examples int he docs of passing in `virtual_text = true` which seemed to be passed directly into `render_virtual_text`, so I thought `true` was a possible value.

```lua
vim.diagnostic.config({ virtual_text = true })
```

```lua
render_virtual_text(
    ns.user_data.virt_text_ns, 
    bufnr, 
    line_diagnostics, 
    opts.virtual_text -- <--- 
)
```

mwcz commented:
> I think a simpler way to implement this would be to check for `opts.current_line == false` in the `render_virtual_text` function. That way we don't have to filter diagnostics yet again, creating another new table. Instead we just skip adding the extmark for the current line if `current_line` is false.

Thanks, that does sound a lot simpler.  I reimplemented it there and it's a lot cleaner.  

If you're happy with the code, I'll update the docs.  Do you think this needs tests?

justinmk commented:
> Do you think this needs tests?

could be a 2-line change in an existing test.

docs https://github.com/neovim/neovim/blob/99e754ae02890d50de13b7bdf3262cfaebe1705c/runtime/lua/vim/diagnostic.lua#L193-L195 should be structured like: 

> * `current_line=true` => only show virtual text on the cursor line
> * `current_line=false` => show virtual text on all lines except the cursor line
> * `current_line=nil` => show virtual text on all lines

mwcz commented:
I updated docs, added a test, applied @gpanders's logic simplification, and squashed up all the commits.  Just two final things.

First, I've been using  the settings this PR enables (`virtual_text = { current_line = false }, virtual_lines = { current_line = true }`) for a while, and it turns out I don't personally like it after all, but I wanted to get this finished up since it seems like some others do want it.

Second, if this is merged, `virtual_text.current_line` and `virtual_lines.current_line` will become inconsistent.  They'll have different default values and behave differently when set to `false`.  Maybe handling for `virtual_lines.current_line = false` should be added as well, but that would be a change purely for consistently, since I don't think anyone would choose that setting.  Thoughts?

justinmk commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -190,8 +190,10 @@ end
 ---
---- Only show diagnostics for the current line.
---- (default `false`)
+--- Show or hide diagnostics based on the current cursor line.  If `true`, only diagnostics on the
+--- current cursor line are shown.  If `false`, all diagnostics are shown except on the current
+--- cursor line.  If `nil`, all diagnostics are shown.
+--- (default `nil`)
```
I mentioned how this should look. It can be much more easily understood as follows. No need for a bunch of prose.

```suggestion
--- Show or hide diagnostics based on the current cursor line.
--- - `true`: only on the cursor line
--- - `false`: on all lines except the cursor line
--- - `nil`: on all lines
---
--- (default `nil`)
```

mwcz replied:
That's pretty close to what I wrote initially.  I then removed the bulleted list and made the wording more conversational because that's the style used throughout most of the other docstrings in diagnostic.lua.  I was aiming for consistency with the existing docs.

justinmk commented:
> Maybe handling for `virtual_lines.current_line = false` should be added as well, but that would be a change purely for consistently, since I don't think anyone would choose that setting. Thoughts?

Yeah, in favor of that.

Meanwhile, the code simplification makes this PR worthwhile in any case. Thank you!
